contracts/utils/cryptography/ERC7913Utils.sol

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.24;
+
+import {SignatureChecker} from "./SignatureChecker.sol";
+import {Bytes} from "../Bytes.sol";
+import {IERC7913SignatureVerifier} from "../../interfaces/IERC7913.sol";
+
+/**
+ * @dev Library that provides common ERC-7913 utility functions.
+ *
+ * This library extends the functionality of xref:api:utils#SignatureChecker[SignatureChecker]
+ * to support signature verification for keys that do not have an Ethereum address of their own
+ * as with ERC-1271.
+ *
+ * See https://eips.ethereum.org/EIPS/eip-7913[ERC-7913].
+ */
+library ERC7913Utils {
+    using Bytes for bytes;
+
+    /**
+     * @dev Verifies a signature for a given signer and hash.
+     *
+     * The signer is a `bytes` object that is the concatenation of an address and optionally a key:
+     * `verifier || key`. A signer must be at least 20 bytes long.
+     *
+     * Verification is done as follows:
+     * - If `signer.length < 20`: verification fails
+     * - If `signer.length == 20`: verification is done using {SignatureChecker}
+     * - Otherwise: verification is done using {IERC7913SignatureVerifier}
+     */
+    function isValidSignatureNow(
+        bytes memory signer,
+        bytes32 hash,
+        bytes memory signature
+    ) internal view returns (bool) {
+        if (signer.length < 20) {
+            return false;
+        } else if (signer.length == 20) {
+            return SignatureChecker.isValidSignatureNow(address(bytes20(signer)), hash, signature);
+        } else {
+            (bool success, bytes memory result) = address(bytes20(signer)).staticcall(
+                abi.encodeCall(IERC7913SignatureVerifier.verify, (signer.slice(20), hash, signature))
+            );
+            return (success &&
+                result.length >= 32 &&
+                abi.decode(result, (bytes32)) == bytes32(IERC7913SignatureVerifier.verify.selector));
+        }
+    }
+
+    /**
+     * @dev Verifies multiple `signatures` for a given hash using a set of `signers`.
+     *
+     * The signers must be ordered by their `keccak256` hash to ensure no duplicates and to optimize
+     * the verification process. The function will return `false` if the signers are not properly ordered.
+     *
+     * Requirements:
+     *
+     * * The `signatures` array must be at least the  `signers` array's length.
+     */
+    function areValidSignaturesNow(
```

Comment:
I'm not a huge fan of having this function here. IMO it should be directly in the MultiSigner. The ordering is not really part of the standard (you could decide to order differently then by hash).

We don't have an equivalent (batch) function in `SignatureChecker`, and I don't think we should have this one here.

Reply:
I proposed these versions a while ago:

- https://github.com/ernestognw/openzeppelin-contracts/blob/f791820e8806912e08214a0266da0314175b5476/contracts/account/modules/TypedERC1271Executor.sol
- https://github.com/ernestognw/openzeppelin-contracts/blob/f791820e8806912e08214a0266da0314175b5476/contracts/account/modules/MultiERC1271Validator.sol

None of them was even taken into consideration because they didn't generalize well in the first place. We also had all that discussion about whether a key has an identity or not.

To me, it's pretty clear now that 1271 didn't generalize well since the beginning. If any, I tried to make the counterpoint that it was enough for generalizing. My opinion is pretty strong on this one